README.md

[![codecov](https://codecov.io/gh/rsc1975/api-mockr/branch/main/graph/badge.svg?token=IHB0J0OOPQ)](https://codecov.io/gh/rsc1975/api-mockr)
[![Code size](https://img.shields.io/github/languages/code-size/rsc1975/api-mockr?logo=github&logoColor=white)](https://github.com/rsc1975/api-mockr)
[![License](https://img.shields.io/github/license/rsc1975/api-mockr?logo=open-source-initiative&logoColor=green)](https://github.com/rsc1975/api-mockr/blob/main/LICENSE)
[![CI](https://img.shields.io/github/workflow/status/rsc1975/api-mockr/Docker%20image?logo=github)](https://github.com/rsc1975/api-mockr/actions/workflows/dockerhub-upload.yml)
[![Last Commit](https://img.shields.io/github/last-commit/rsc1975/api-mockr?logo=github)](https://github.com/rsc1975/api-mockr/commits/main)
[![Docker version](https://img.shields.io/docker/v/dvlprtech/api-mockr?sort=semver&logo=docker)](https://hub.docker.com/r/dvlprtech/api-mockr)
[![NPM version](https://img.shields.io/npm/v/api-mockr?logo=npm)](https://www.npmjs.com/package/api-mockr)


# api-mockr

Just another API mocker that can be configured to response random and specific data. It can help testing integrations with third parties. It's specially useful to create integration tests where we need to interact with an external service, so we can customize the response for different services exactly as we expect the external service responses. 

Currently, only REST API responses are supported.


# Demo

There is a demo available [here](https://api-mockr.dvlpr.tech/). The application is deployed in the free tier of [Cyclic](https://cyclic.sh/), so please, don't abuse.
# Getting started

This is a NodeJS application, you can install it as dev dependency in your project or as independent project, there is also a [Docker Image](https://hub.docker.com/r/dvlprtech/api-mockr).

## Installation

Installation as dev dependency:

```sh
npm install --save-dev api-mockr
```	

Installation as docker image:

```sh
docker pull dvlprtech/api-mockr
```

For further details about the use with docker you can access to DockerHub [api-mockr image](https://hub.docker.com/r/dvlprtech/api-mockr) page.

## How to use

You can use the CLI tool `api-mockr` to start the application:

```sh
npx api-mockr
```

The following output will apear:

```sh
[🟢 api-mockr v0.0.0] Server running at: http://0.0.0.0:3003
```

There are several options params that can be used to configure the server:

* `--port`: The port to listen to. Default is `3003`.
* `--host`: The host to listen to. Default is `0.0.0.0`.
* `--silent`: If present, the server will not output any message. By default, the server will print every request with the basic info about the response.
* `--verbose`: If present, the server will print every request with its params and payload and the applied server configuration when the server is created.
* `--config`: The path to the config file. It's a multivalue param.
* `--apiPrefix`: The prefix to invoque the API services. Default is `/`. If set all mocked routes should be invoked with this prefix. I.e. The configured route `/v1/users` should be invoked as `/api/v1/users` if we use `--apiPrefix=/api`.

Example of use:
    
```sh
$ npx api-mockr --port=13003 --host=localhost --apiPrefix=/api
[🟢 api-mockr v0.0.0] Server running at: http://localhost:13003/api
```

Without additional configuration, the default server will response with the following data to whatever request:

```txt
$ curl 'http://localhost:3003/api/whatever?foo=bar'
{"success":true,"request":{"path":"/api/whatever","params":{"foo":"bar"}}}
```

There is an exception, the call to `/` will return a fixed text, that can be used to "ping" the server.
```txt
$ curl 'http://localhost:3003/'
API Mockr (v0.0.0)
```

## Special params

There are some special params that can be used with each request:

* `_pretty`: If it's set to `true`, the response will be pretty printed.
* `_delay`: The delay in milliseconds to wait before responding. By default 0.
* `_forceError`: If it's set to `true`, the response will be an error with a http status code of `500` by default.
* `_clonePayload`: If it's set to `true`, the response payload will be a clone of the request payload but appling variable substitution if they are used in request payload.


The error response can be customized with 2 header:

* `x-mocker-force-error'`: HTTP status code to be returned, can be any value between `400` and `599`. Values out of range will be ignored.
* `x-mocker-error-msg`: The error message to be returned, you can use variables in the message, like `"ERROR: ${random.emoji} ${random.phrase}"`.

Example of error response:

```txt
$ curl 'http://localhost:3003/api/whatever?foo=bar&_forceError=true&_pretty=1'
{
  "success": false,
  "error": "Error in request to path: /api/whatever",
  "request": {
    "body": null,
    "params": {
      "foo": "bar",
      "_forceError": "true",
      "_pretty": "1"
    }
  }
}
```

Customizing the message and code:

```txt
$ curl -H "x-mocker-force-error: 418" -H 'x-mocker-error-msg: ERROR: ${random.emoji} ${random.phrase}' 'http://localhost:3003/api/whatever?foo=bar&_forceError=true&_pretty=1'
{
  "success": false,
  "error": "ERROR: 🐭 You cant quantify the driver without transmitting the multi-byte SQL microchip!",
  "request": {
    "body": null,
    "params": {
      "foo": "bar",
      "_forceError": "true",
      "_pretty": "1"
    }
  }
```

The emoji and the error phrase will be different with each call.

## Server configuration

The config files can be read in YAML and JSON formats, the default config file is like this:

```yaml
defaultResponse:
  success: true
  timestamp: ${server.isoDatetime}
  request:
    path: ${request.path} 
    body: ${request.payload}
    params: ${request.params}
errorResponse:
  $httpStatus$: 500
  success: false
  error: "${error}"
  request:
    body: ${request.payload}
    params: ${request.params}
routes:
  get: 
    "say-whatever":
      success: true 
      author: ${random.personFirstName}
      phrase: ${random.phrase}
```

There are 3 main sections in the config file:

* `defaultResponse`: The default response to be returned when there is no match for the request path and the configured routes
* `errorResponse`: The error response payload to be returned when an error occurs.
* `routes`: The routes to be mocked grouped by HTTP method (get, post, put, ...). A wildcard (`*`) can be used to match any method and/or any path.

## Variables in configuration

We can use variables in the configuration files to provide dynamic values with every response. There are several variables types:

* `random`: Random values for all sort of data. i.e. `random.personFullName`, `random.integer`, `random.department`, ...
* `request`: Values from the request. i.e. `request.path`, `request.params`, `request.payload`
* `server`: Values from the server. i.e. `server.timestamp`, `server.isoDate`, ...
* Path variables: Variables created in the config file to refer a path section. i.e. A configured route like `/api/users/${username}` will create a variable `username` with value "bob23" for a request path like `/api/users/bob23`, this variable can be used in the route response (`... "author": "${username}", ...`).

Some of the variables accept parameters, the params should be added after a dot (`variable.paramValue`) for example:

* `request.params.foo`: In requests like `/api/whatever?foo=bar` It will return `bar`.
* `random.personFirstName.male`: This params accepts `male` or `female` as possible input values
* `random.choose.monday.sunday.tuesday`: This params accepts a variable number of params and It will return one of them randomly.

For further informaton about the supported variables, please refer to the [VARS.md](https://github.com/rsc1975/api-mockr/blob/main/VARS.md) file.

There are several examples of config files in [examples directory](./examples).



CHANGELOG.md
## 1.0.0

- Initial version.

## 0.9.14

- Initial beta version.






src/mock_server.ts
import { LogEvent, Request, ResponseToolkit, server as hapiServer, Server } from '@hapi/hapi';
import { ResponseGenerator } from './response_generator';
import { MockerConfig } from './routes_config';

interface MockServerInputParams {
    version?: string;
    port?: number;
    host?: string;
    verbose?: boolean;
    silent?: boolean;
    responseConfig?: MockerConfig;   
    apiPrefix?: string;
}

const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

export const PING_MSG = 'API Mockr';
const PRETTY_PARAM = '_pretty';
const DELAY_PARAM = '_delay';
const FORCE_ERROR_PARAM = '_forceError';
const FORCE_ERROR_HEADER = 'x-mocker-force-error';
const FORCE_ERROR_MSG_HEADER = 'x-mocker-error-msg';

export class MockServer {
    public server: Server;
    public version: string;
    public host: string;
    public port: number;
    public apiPrefix: string;
    public verbose: boolean;
    public silent: boolean;
    public responseConfig: MockerConfig;

    constructor({host, port, apiPrefix, verbose, silent, responseConfig, version} : MockServerInputParams = {}) {
        this.apiPrefix = apiPrefix || process.env.MOCKER_PREFIX || '';
        if (!this.apiPrefix.startsWith("/") && !!this.apiPrefix) {
            this.apiPrefix = "/" + this.apiPrefix;
          }
        this.version = version || 'unknown';
        this.host = host || process.env.MOCKER_BINDING || '0.0.0.0';
        this.port = port || +(process.env.MOCKER_PORT || 3003);
        this.silent = !!silent;
        this.verbose = !this.silent && !!verbose;
        this.responseConfig = responseConfig || {};
        this.server = this.create();
        if (this.verbose) {
            this.printConfig(this.responseConfig);
        }
    }

    private create() : Server {
        const srv =  hapiServer({
            port: this.port,
            host: this.host,
            routes: {
                cors: true,
                payload: {
                    output: 'data',
                    parse: true,                    
                }
            }
        });
        srv.route({
            method: 'GET',
            path: '/',
            handler: (_, h) => h.response(`${PING_MSG} (v${this.version})`).header('content-type', 'text/plain').takeover()
        });
    
        // Create hapi server catchall route
        srv.route({
            method: '*',
            path: `${this.apiPrefix}/{p*}`,
            handler: (req: Request, h : ResponseToolkit) => {
                const responseGenerator = new ResponseGenerator(this.responseConfig, this.apiPrefix);
                return responseGenerator.generate(req);
            }
        });

        const logRequest = (req: Request) => {
            
            let logReq = `[DEBUG] [➡️] ${req.method.toUpperCase()} ${req.url.pathname}`;
            if (!!req.url.search) {
                logReq += ` [params: ${req.url.search.substring(1).replace('&', ' ')}]`;
            }
            if (!!req.payload) {
                const body = req.headers['content-type'] === 'application/json' ? JSON.stringify(req.payload) : req.payload.toString();
                
                logReq += ` [payload: ${body}]`;
            }
            
            this.server.log('verbose', logReq);
        }

        srv.ext('onPreHandler', (req : Request, h : ResponseToolkit) => {
            if (this.verbose) {
                logRequest(req);
            }
            return h.continue;
        });

        srv.ext('onRequest', async (req : Request, h : ResponseToolkit) => {
            const { headers } : any = req;
            const delay = +req.query[DELAY_PARAM];
            if (delay) {
                await sleep(delay);
            }
            return h.continue;
        });

        srv.ext('onPreResponse', (req : Request, h : ResponseToolkit) => {
            const { response, headers } : any = req;
            if (response.output?.statusCode === 404) {
                return h.response(`Missing route, try: ${this.apiPrefix}/<anything>`).code(404).takeover();
            }
            if (req.query[PRETTY_PARAM] && typeof response.source === 'object') {
                response.spaces(2).takeover();
            }
            const forceError = !!req.query[FORCE_ERROR_PARAM];
            const errorCode = +headers[FORCE_ERROR_HEADER];
            if (!!errorCode || forceError) {                
                const errorMsg = headers[FORCE_ERROR_MSG_HEADER];
                const responseGenerator = new ResponseGenerator(this.responseConfig, this.apiPrefix);
                const errorResponse = responseGenerator.generateError(req, errorMsg, errorCode);
                return h.response(errorResponse.payload).code(errorResponse.httpStatus).takeover();
            }
            return h.continue;
        });
        srv.ext('onPostResponse', (req : Request, h : ResponseToolkit) => {
            const { response } : any = req;
            this.server.log('info', `[INFO] [➡️] ${req.method.toUpperCase()} ${req.url.pathname} [↩️ ${response.statusCode >= 400 ? '⭕' : '✅'}] (status: ${response.statusCode}, length: ${response.headers['content-length'] }, content-type: ${response.contentType})`);
            return h.continue;
        });        

        srv.events.on('log', (event: LogEvent, tags: { [key: string]: true }) => {
            if (!this.silent) {
                if (tags.info) {
                    console.info(event.data);
                }
                if (this.verbose && tags.verbose) {
                    console.info(event.data);
                }
            }
        });
    
        
        return srv;
    }


    async start() : Promise<void> {       
        await this.server.initialize();        
        this.server.log('info', `[🟢 api-mockr v${this.version}] Server running at: ${this.server.info.uri}${this.apiPrefix}`);
        return this.server.start();
    }

    async stop() {
        await this.server.stop();
        this.server.log('info', `[🟥 api-mockr] Server stopped`);
    }

    async dispose() {
        await this.stop();
    }

    printConfig(config: MockerConfig) {
        this.server.log('verbose', "[INFO] API Mockr is using the following configuration:");
        this.server.log('verbose', JSON.stringify(config, null, 2)+"\n");        
    }
    
}
